core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageInputStreamTest.java

```diff
@@ -0,0 +1,315 @@
+    }
+
+    @Test
+    void readWithStreamMessageOperators() throws Exception {
+        final StreamMessage<Integer> streamMessage = StreamMessage
+                .of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
+                .filter(x -> x % 2 == 0)
+                .map(x -> x + 10); // 12, 14, 16, 18, 20
+        final InputStream inputStream = streamMessage
+                .asInputStream(x -> HttpData.wrap(x.toString().getBytes()));
+        final byte[] expected = ImmutableList.of(12, 14, 16, 18, 20)
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void readWithOffset() throws Exception {
+        final StreamMessage<byte[]> streamMessage = StreamMessage.of(new byte[] {1, 2, 3, 4, 5});
+        final InputStream inputStream = streamMessage.asInputStream(HttpData::wrap);
+
+        final byte[] result = new byte[5];
+        final int len = inputStream.read(result, 1, 3);
+
+        assertThat(len).isEqualTo(3);
+        assertThat(result).isEqualTo(new byte[] { 0, 1, 2, 3, 0});
+        assertThat(inputStream.available()).isEqualTo(2);
+    }
+
+    @Test
+    void readByMultipleThreads_allThreadsShouldWaitSubscribed() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(10);
+        final Publisher<Integer> publisher = Flux
+                .range(1, 10)
+                .doOnSubscribe(subscription -> {
+                    try {
+                        latch.await(); // to ensure that all threads are waiting onSubscribe()
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                });
+        final StreamMessage<Integer> streamMessage = new PublisherBasedStreamMessage<>(publisher);
+        final InputStream inputStream = streamMessage
+                .asInputStream(x -> HttpData.wrap(x.toString().getBytes()));
+
+        final List<Thread> threads = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            final Thread t = new Thread(() -> assertDoesNotThrow(() -> {
+                latch.countDown();
+                inputStream.read();
+            }));
+            threads.add(t);
+            t.start();
+        }
+        for (Thread thread : threads) {
+            thread.join();
+        }
+    }
```

Comment:
```java
    private void ensureSubscribed() {
        if (subscribed.compareAndSet(false, true)) {
            source.subscribe(subscriber);
        }
        subscriber.whenSubscribed.join();
    }
```
- I added `readByMultipleThreads` test to check `ensureSubscribed()` on `read` by `multithread`

```java
    private void ensureSubscribed() {
        if (subscribed.compareAndSet(false, true)) {
            source.subscribe(subscriber);
            // only first read() thread will wait onSubscribe()..
            subscriber.whenSubscribed.join();
        }
    }
```
- If `whenSubscribed.join()` like above, `upstream` can be `null` by timing issue on `nextStream()`! 
<img width="1395" alt="스크린샷 2022-03-05 오후 11 20 04" src="https://user-images.githubusercontent.com/34854527/156888172-38eae7af-2fa1-4e43-a124-843a91c6fba2.png">

- (checked on my local env)

Reply:
As a stream is a chain of objects, I don't think we need to support reading for multiple threads and thread safety.
The stream may need to be read by a single thread to preserve the order of data.

In threading environments, the following blocking should also be atomic.
https://github.com/line/armeria/blob/57b61938b3b773fbf1e2656aa4fd9f6b82643662/core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java#L85-L87

For `StreamMessageInputStream`, we should consider two threads. one is for writer/subscription thread and another one is reader thread.

Reply:
Additionally, we need a `volatile` keyword to make `upstream` visible to the reader thread and avoid NPE.

Reply:
> For StreamMessageInputStream, we should consider two threads. one is for writer/subscription thread and another one is reader thread.

aha~ I understood your intention about "multiple thread" here!

> Additionally, we need a volatile keyword to make upstream visible to the reader thread

✅ updated, I miss this case. thanks!